## Crime Time - Acquisition

- Source data is from https://data.austintexas.gov/Public-Safety/Annual-Crime-Dataset-2015/spbg-9v94
    
- Public data that was exported originally to Apple Numbers CSV

### Imports

- standard at the beginning; will adjust as we need them

In [1]:
# import warnings
# warnings.simplefilter('ignore')

import pandas as pd
import numpy as np
import os

import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# pull in the data from csv 

# format:
#  dataframe = pd.read_csv(r"path.csv")

df = pd.read_csv (r"/Users/DataScience/Desktop/Annual_Crime_Dataset_2015.csv")

df


GO Primary Key  Council District         GO Highest Offense Desc  \
0           201510782               4.0  AGG ROBBERY/DEADLY WEAPON        
1           201511231               4.0  ROBBERY BY ASSAULT               
2           201511736               1.0  BURGLARY OF RESIDENCE            
3           201511433               4.0  BURGLARY OF RESIDENCE            
4           201511936               2.0  BURGLARY OF RESIDENCE            
...               ...               ...                             ...   
38568     20153650803               1.0  THEFT                            
38569     20155056956               6.0  THEFT                            
38570     20153651059              10.0  THEFT                            
38571     20153651185               1.0  THEFT                            
38572     20155056975               7.0  THEFT                            

      Highest NIBRS/UCR Offense Description GO Report Date  \
0                                   Robbery       1-Jan-15   
1                                   Robbery       1-Jan-15   
2                                  Burglary       1-Jan-15   
3                                  Burglary       1-Jan-15   
4                                  Burglary       1-Jan-15   
...                                     ...            ...   
38568                                 Theft      31-Dec-15   
38569                                 Theft      31-Dec-15   
38570                                 Theft      31-Dec-15   
38571                                 Theft      31-Dec-15   
38572                                 Theft      31-Dec-15   

                      GO Location Clearance Status Clearance Date GO District  \
0            9001 N IH 35 SVRD NB                N      28-Jan-15           E   
1         919 E KOENIG LN SVRD EB                N      13-Jan-15           I   
2           12151 N IH 35 SVRD NB                N      13-Jan-15           E   
3          1044 NORWOOD PARK BLVD                N       5-Jan-15           I   
4            2413 BITTER CREEK DR                N       7-Jan-15           F   
...                           ...              ...            ...         ...   
38568      1100 BLOCK HARGRAVE ST              NaN            NaN           C   
38569  13776 N US 183 HWY SVRD SB                N      13-Jan-16           A   
38570      7025 VILLAGE CENTER DR                C       8-Feb-16           B   
38571              6575 DECKER LN                N      10-Feb-16           C   
38572           12113 METRIC BLVD                N      14-Jan-16           E   

       GO Location Zip  GO Census Tract  GO X Coordinate  GO Y Coordinate  
0              78753.0            18.13        3130483.0       10102366.0  
1              78751.0            21.05        3124730.0       10090296.0  
2              78753.0            18.35        3135985.0       10117220.0  
3              78753.0            18.13        3129896.0       10096032.0  
4              78744.0            24.27        3110455.0       10039340.0  
...                ...              ...              ...              ...  
38568          78702.0             8.02        3124675.0       10072478.0  
38569          78750.0           204.03        3096285.0       10140738.0  
38570          78731.0            17.18        3109504.0       10101310.0  
38571          78724.0            22.11        3150290.0       10079608.0  
38572          78758.0            18.45        3127286.0       10120469.0  

[38573 rows x 13 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38573 entries, 0 to 38572
Data columns (total 13 columns):
GO Primary Key                           38573 non-null int64
Council District                         38401 non-null float64
GO Highest Offense Desc                  38573 non-null object
Highest NIBRS/UCR Offense Description    38573 non-null object
GO Report Date                           38573 non-null object
GO Location                              38092 non-null object
Clearance Status                         36649 non-null object
Clearance Date                           36649 non-null object
GO District                              38573 non-null object
GO Location Zip                          38414 non-null float64
GO Census Tract                          38414 non-null float64
GO X Coordinate                          37981 non-null float64
GO Y Coordinate                          37981 non-null float64
dtypes: float64(5), int64(1), object(7)
memory usage: 3.8+ MB


In [4]:
# Lotsa 'nulls' it seems...

print("\nCount total NaN at each column in a DataFrame : \n \n", 
      df.isnull().sum())


Count total NaN at each column in a DataFrame : 
 
 GO Primary Key                              0
Council District                          172
GO Highest Offense Desc                     0
Highest NIBRS/UCR Offense Description       0
GO Report Date                              0
GO Location                               481
Clearance Status                         1924
Clearance Date                           1924
GO District                                 0
GO Location Zip                           159
GO Census Tract                           159
GO X Coordinate                           592
GO Y Coordinate                           592
dtype: int64


#### Quick glance of the data shows us we're dealing with a lot of nulls, particularly in the following columns:

- 172 in 'Council District' (~ 0.44%)

- 481 in 'GO Location' (~ 1.24%)

- 1,924 in BOTH 'Clearance Status' and 'Clearance Date' (~ 5.0%)

- 159 in BOTH 'GO Location Zip' and 'GO Census Tract' (~ 0.41%)

- 592 in BOTH 'GO X Coordinate' and 'GO Y Coordinate' (~ 1.5%)

> #### Also, we'll have to change the datatypes in the following columns from strings to floats / ints:

    - 'GO Highest Offense Desc'
    - 'Highest NIBRS / UCR Offense Description'
    - 'GO Report Date'
    - 'GO Location'
    - 'Clearance Status'
    - 'Clearance Date'
    - 'GO District'

In [5]:
# Taking a look at the unique values 

print(df.apply(pd.Series.nunique, axis=0))

GO Primary Key                           38573
Council District                            10
GO Highest Offense Desc                     44
Highest NIBRS/UCR Offense Description        7
GO Report Date                             365
GO Location                              16764
Clearance Status                             3
Clearance Date                             433
GO District                                 11
GO Location Zip                             47
GO Census Tract                            205
GO X Coordinate                          14206
GO Y Coordinate                          15187
dtype: int64


**Seeing if the 'df.apply' function worked by checking individual counts:**

In [6]:
print("The unique values in 'GO Primary Key' are: ")
print(df["GO Primary Key"].nunique())
print("The unique values in 'Council District' are: ")
print(df["Council District"].nunique())
print("The unique values in 'GO Highest Offense Desc' are: ")
print(df["GO Highest Offense Desc"].nunique())
print("The unique values in 'GO Location' are: ")
print(df["GO Location"].nunique())
print("The unique values in 'GO Location Zip' are: ")
print(df["GO Location Zip"].nunique())

The unique values in 'GO Primary Key' are: 
38573
The unique values in 'Council District' are: 
10
The unique values in 'GO Highest Offense Desc' are: 
44
The unique values in 'GO Location' are: 
16764
The unique values in 'GO Location Zip' are: 
47


**^^Checks out^^**

**Find out where the NaN values are in each of the columns to determine whether or not they should be kept.**

In [7]:
df[df.isnull().any(axis=1)]

GO Primary Key  Council District         GO Highest Offense Desc  \
34          201511491               4.0  THEFT BY SHOPLIFTING             
50         2015560616               4.0  RAPE                             
60         2015520845               2.0  RAPE                             
84          201520100               1.0  AGG ROBBERY/DEADLY WEAPON        
96          201520326               4.0  BURGLARY OF RESIDENCE            
...               ...               ...                             ...   
38531     20155056981               7.0  BURGLARY OF VEHICLE              
38538     20155057000               3.0  BURGLARY OF VEHICLE              
38556     20158004696               7.0  THEFT                            
38563     20155056947               8.0  THEFT                            
38568     20153650803               1.0  THEFT                            

      Highest NIBRS/UCR Offense Description GO Report Date  \
34                                    Theft       1-Jan-15   
50                                     Rape      25-Feb-15   
60                                     Rape      21-Feb-15   
84                                  Robbery       2-Jan-15   
96                                 Burglary       2-Jan-15   
...                                     ...            ...   
38531                                 Theft      31-Dec-15   
38538                                 Theft      31-Dec-15   
38556                                 Theft      31-Dec-15   
38563                                 Theft      31-Dec-15   
38568                                 Theft      31-Dec-15   

                      GO Location Clearance Status Clearance Date GO District  \
34              6001 AIRPORT BLVD              NaN            NaN           I   
50                            NaN                O      22-Apr-15           E   
60                            NaN                O      11-Mar-15           F   
84                1209 DELONEY ST              NaN            NaN           C   
96                9844 QUAIL BLVD              NaN            NaN           E   
...                           ...              ...            ...         ...   
38531  10706 N MOPAC EXPY SVRD SB              NaN            NaN           A   
38538            2000 WOODWARD ST              NaN            NaN           H   
38556       8900 SHOAL CREEK BLVD              NaN            NaN           B   
38563    4625 W WILLIAM CANNON DR              NaN            NaN           D   
38568      1100 BLOCK HARGRAVE ST              NaN            NaN           C   

       GO Location Zip  GO Census Tract  GO X Coordinate  GO Y Coordinate  
34             78752.0            15.03        3122890.0       10091831.0  
50             78758.0            18.05              NaN              NaN  
60             78744.0            24.27              NaN              NaN  
84             78721.0            21.09        3130038.0       10075027.0  
96             78758.0            18.22        3126613.0       10108839.0  
...                ...              ...              ...              ...  
38531          78759.0            17.54        3116246.0       10117603.0  
38538          78741.0            23.07        3114537.0       10052722.0  
38556          78757.0            18.17        3116314.0       10109976.0  
38563          78749.0            17.38        3082868.0       10053773.0  
38568          78702.0             8.02        3124675.0       10072478.0  

[2483 rows x 13 columns]

### Data Dictionary

- 2,483 rows with NaN values.  That's roughly 6.4% of all the total information.  That seems like a big chunk, so let's investigate the columns and what they mean:

>*NOTE:* Throughout the dataset, you will see the abbreviation 'GO,' which appears to be short for either 'General Offense' or 'General Occurence.'  There's no indication as to its definition within the dataset documentation, and the 'general offense' description is based on 'GO' as it is used by the Minneapolis, MN police department (found via reddit).

**GO Primary Key -** Based on the fact that there are 38,573 rows of data and 38,573 unique values for this column, it is safe to say that the 'GO Primary Key' is the unique identifier for each incident.

**Council District -** The district number in which the event occurred.  A 'NaN' value here means the district number was not documented.  Seems like we would keep it, but with less than half a percent of missing values, we can **drop the NaNs** and still maintain enough data to give an accurate depiction of the crime rate.

**GO Highest Offese Desc -** A more detailed description of the 'NIBRS/UCR' column that follows.  Where the 'NIBRS/UCR' is a general category of crime, the 'GO Highest Offense Desc' is a more detailed explanation of the reported event.  

**Highest NIBRS/UCR Offense Description -** 'U.C.R' is short for 'Uniform Crime Reporting,' and 'N.I.B.R.S.' is the acronym for the National Incident-Based Reporting System, as described by the [FBI's website](https://ucr.fbi.gov/nibrs/2012/resources/nibrs-offense-definitions).  The NIBRS is used for general categorization of crimes based on common-law definitions found in 'Black's Law Dictionary,' the most widely-used law dictionary in the United States, and a secondary legal authority to the U.S. Supreme Court.

**GO Location -** Incident location, and it looks to be the street address where the crime occurred.  According to the website, rape locations are not listed so as to protect the id of the victim.  A quick glance at the data shows that the first NaNs we find are indeed rapes; that's a crime we don't want to eliminate from investigation, so I am **not dropping** these rows.  Instead, we can impute a value where there are NaNs.

**Clearance Status -** Used to identify whether or not a crime was closed.  

- 'Cleared by Arrest' means at least one person has been arrested, charged, AND handed over to the court for prosecution

- 'Cleared by Exception' means the offender has been identified, evidence of their involvement has been collected and given to the court, and the suspect's location for arrest is known, BUT the police could not arrest him / her because something out of the police's control happened (suspect is dead or can't be extradited, for example) 

>**C** = 'Cleared by Arrest' | 
>**O** = 'Cleared by Exception' |
>**N** = 'Not Cleared'

- nearly 5% of these values are 'NaN,' one, and two: they still list the crime, so **not dropping**.

**Clearance Date -** Seems to correspond directly with 'Clearance Status'; ie: the two column values match. Therefore, **will not drop** because dates are important in determining seasonality of events.  BUT need to **set these dates to datetime** for better time series analysis (*same must be done for 'GO Report Date' column*).  

**GO Location Zip -** Corresponds with the 'GO Location' street addresses, but IS included where there are 'NaN' values on street addresses with respect to rape cases.  With less than a half-percent of its fields missing values, we will **drop**.

**GO Census Tract -** Neighborhoods.  No clear info given from the website, but according to the [US Census Beureau](https://www.census.gov/programs-surveys/geography/about/glossary.html#par_textimage_13), it's a geographic area with somewhere between 1,200 and 8,000 people.  While the glossary goes on into what the numbers and their suffixes mean, this brief description satisfies our needs.  And while less than a half-percent of the values are listed as 'NaN,' we can **drop** these rows.

**GO X Coordinate -** The geospacial x-coordinate for the corresponding crime scene location, per NAD 1983 State Plane Texas Central FIPS 4203 (US Feet), or longitudinal / latitudinal coordinates.  Given that we already have a street address and zipcode, we can not only **drop** theses 'NaN' rows, we can **drop the entire column**.

**GO Y Coordinate -**  Similar to the 'GO X Coordinate,' we can **drop the entire column**.

In [8]:
# dropping 'GO X / Y Coordinate' columns

df = df.drop(columns = ["GO X Coordinate", "GO Y Coordinate"])

In [9]:
df.head()

GO Primary Key  Council District         GO Highest Offense Desc  \
0       201510782               4.0  AGG ROBBERY/DEADLY WEAPON        
1       201511231               4.0  ROBBERY BY ASSAULT               
2       201511736               1.0  BURGLARY OF RESIDENCE            
3       201511433               4.0  BURGLARY OF RESIDENCE            
4       201511936               2.0  BURGLARY OF RESIDENCE            

  Highest NIBRS/UCR Offense Description GO Report Date  \
0                               Robbery       1-Jan-15   
1                               Robbery       1-Jan-15   
2                              Burglary       1-Jan-15   
3                              Burglary       1-Jan-15   
4                              Burglary       1-Jan-15   

               GO Location Clearance Status Clearance Date GO District  \
0     9001 N IH 35 SVRD NB                N      28-Jan-15           E   
1  919 E KOENIG LN SVRD EB                N      13-Jan-15           I   
2    12151 N IH 35 SVRD NB                N      13-Jan-15           E   
3   1044 NORWOOD PARK BLVD                N       5-Jan-15           I   
4     2413 BITTER CREEK DR                N       7-Jan-15           F   

   GO Location Zip  GO Census Tract  
0          78753.0            18.13  
1          78751.0            21.05  
2          78753.0            18.35  
3          78753.0            18.13  
4          78744.0            24.27

**Column names are wordy, and 'GO' is repetitive.  So: renaming columns for ease of reading**

In [15]:
df = df.rename(
            columns={
                "GO Primary Key": "incident_id",
                "Council District": "district_council_num",
                "GO Highest Offense Desc": "detailed_description",
                "Highest NIBRS/UCR Offense Description": "general_category",
                "GO Report Date": "incident_report_date",
                "GO Location": "incident_address",
                "Clearance Status": "clearance_status",
                "Clearance Date": "clearance_date",
                "GO District": "district_letter",
                "GO Location Zip": "incident_zipcode",
                "GO Census Tract": "neighborhood",
            }
        )

df # print df to show renamed columns

incident_id  district_council_num            detailed_description  \
0        201510782                   4.0  AGG ROBBERY/DEADLY WEAPON        
1        201511231                   4.0  ROBBERY BY ASSAULT               
2        201511736                   1.0  BURGLARY OF RESIDENCE            
3        201511433                   4.0  BURGLARY OF RESIDENCE            
4        201511936                   2.0  BURGLARY OF RESIDENCE            
...            ...                   ...                             ...   
38568  20153650803                   1.0  THEFT                            
38569  20155056956                   6.0  THEFT                            
38570  20153651059                  10.0  THEFT                            
38571  20153651185                   1.0  THEFT                            
38572  20155056975                   7.0  THEFT                            

      general_category incident_report_date            incident_address  \
0              Robbery             1-Jan-15        9001 N IH 35 SVRD NB   
1              Robbery             1-Jan-15     919 E KOENIG LN SVRD EB   
2             Burglary             1-Jan-15       12151 N IH 35 SVRD NB   
3             Burglary             1-Jan-15      1044 NORWOOD PARK BLVD   
4             Burglary             1-Jan-15        2413 BITTER CREEK DR   
...                ...                  ...                         ...   
38568            Theft            31-Dec-15      1100 BLOCK HARGRAVE ST   
38569            Theft            31-Dec-15  13776 N US 183 HWY SVRD SB   
38570            Theft            31-Dec-15      7025 VILLAGE CENTER DR   
38571            Theft            31-Dec-15              6575 DECKER LN   
38572            Theft            31-Dec-15           12113 METRIC BLVD   

      clearance_status clearance_date district_letter  incident_zipcode  \
0                    N      28-Jan-15               E           78753.0   
1                    N      13-Jan-15               I           78751.0   
2                    N      13-Jan-15               E           78753.0   
3                    N       5-Jan-15               I           78753.0   
4                    N       7-Jan-15               F           78744.0   
...                ...            ...             ...               ...   
38568              NaN            NaN               C           78702.0   
38569                N      13-Jan-16               A           78750.0   
38570                C       8-Feb-16               B           78731.0   
38571                N      10-Feb-16               C           78724.0   
38572                N      14-Jan-16               E           78758.0   

       neighborhood  
0             18.13  
1             21.05  
2             18.35  
3             18.13  
4             24.27  
...             ...  
38568          8.02  
38569        204.03  
38570         17.18  
38571         22.11  
38572         18.45  

[38573 rows x 11 columns]

In [12]:
# setting 'GO Report Date' and 'Clearance Date' values from dates to datetime

df["GO Report Date"] = pd.to_datetime(df["GO Report Date"])
df["Clearance Date"] = pd.to_datetime(df["Clearance Date"])

KeyError: 'GO Report Date'

In [ ]:
# Sort rows by the report date and make the index that date

df = df.sort_values("GO Report Date").set_index("GO Report Date")

df.head(10)

**So, what does the info above tell us?**

- multiple events happen every day - each row is NOT a single day, but a single, specific event

- how 'bout that 'Theft,' huh?  Yikes.  Let's take a look at just how prevalent 'theft' is using the NIBRS category:

In [ ]:
# checking how many of the overall crimes are categorized under NIBRS as 'theft'
(df.Highest NIBRS/UCR Offence Description == "THEFT").sum()

**Let's now take a look at the crime by the quarter, using date ranges Jan - Mar, Apr - June, July - Sept, Oct - Dec (thus showcasing the beauty of datetimes as indeces):**

In [ ]:
# Return all rows from First Quarter, January 1 - March 31 

df.loc["2015-01-01":"2015-03-31"]

In [ ]:
# Return all rows from Second Quarter, April 1 - June 30

df.loc["2015-04-01":"2015-06-30"] 

In [ ]:
# Return all rows from Third Quarter, July 1 - September 30

df.loc["2015-07-01":"2015-09-30"] 

In [ ]:
# Return all rows from Fourth Quarter, Oct 1 - December 31

df.loc["2015-10-01":"2015-12-31"] 

**Since the number of rows == the number of events, let's take a quick look at the data to see in which quarter most of the reported crimes took place:**

In [ ]:
q1 = df.loc["2015-01-01":"2015-03-31"]
q2 = df.loc["2015-04-01":"2015-06-30"] 
q3 = df.loc["2015-07-01":"2015-09-30"] 
q4 = df.loc["2015-10-01":"2015-12-31"]

print(f"First Quarter had {q1.shape[0]} reported events take place")
print(f"Second Quarter had {q2.shape[0]} reported events take place")
print(f"Third Quarter had {q3.shape[0]} reported events take place")
print(f"Fourth Quarter had {q4.shape[0]} reported events take place")

**^^Clearly, Third Quarter had the most reported events, followed by Quarters Two, Four, and One.**

In [ ]:

df2 = pd.DataFrame({'lab':['A', 'B', 'C'], 'val':[10, 30, 20]})

print(df2)

ax = df2.plot.bar(x='lab', y='val', rot=0)

# ax = df.plot.bar(x="GO Report Date") 

# #, y='val', rot=0)

In [ ]:
# imputing values for NaN in 'GO Location:'

from sklearn.impute import SimpleImputer

In [ ]:
imputer = SimpleImputer(missing_values = np.nan, strategy = "constant")